src/dawn/engine_dawn.cc

```diff
@@ -179,8 +189,253 @@ MapResult MapBuffer(const ::dawn::Device& device, const ::dawn::Buffer& buf) {
+  constexpr ::dawn::ShaderStageBit kNoStages{};
+
+  std::vector<::dawn::BindGroupLayoutBinding> bindings;
+  for (const ::dawn::BindGroupLayoutBinding& binding : bindingsInitializer) {
+    if (binding.visibility != kNoStages) {
+      bindings.push_back(binding);
+    }
+  }
+
+  ::dawn::BindGroupLayoutDescriptor descriptor;
+  descriptor.bindingCount = static_cast<uint32_t>(bindings.size());
+  descriptor.bindings = bindings.data();
+  return device.CreateBindGroupLayout(&descriptor);
+}
+
+::dawn::PipelineLayout MakeBasicPipelineLayout(
+    const ::dawn::Device& device,
+    const ::dawn::BindGroupLayout* bindGroupLayout) {
+  ::dawn::PipelineLayoutDescriptor descriptor;
+  if (bindGroupLayout) {
+    descriptor.bindGroupLayoutCount = 1;
+    descriptor.bindGroupLayouts = bindGroupLayout;
+  } else {
+    descriptor.bindGroupLayoutCount = 0;
+    descriptor.bindGroupLayouts = nullptr;
+  }
+  return device.CreatePipelineLayout(&descriptor);
+}
+
+::dawn::TextureView CreateDefaultDepthStencilView(
+    const ::dawn::Device& device,
+    const ::dawn::TextureFormat depth_stencil_format) {
+  ::dawn::TextureDescriptor descriptor;
+  descriptor.dimension = ::dawn::TextureDimension::e2D;
+  descriptor.size.width = 250;
+  descriptor.size.height = 250;
+  descriptor.size.depth = 1;
+  descriptor.arrayLayerCount = 1;
+  descriptor.sampleCount = 1;
+  descriptor.format = depth_stencil_format;
+  descriptor.mipLevelCount = 1;
+  descriptor.usage = ::dawn::TextureUsageBit::OutputAttachment;
+  auto depthStencilTexture = device.CreateTexture(&descriptor);
+  return depthStencilTexture.CreateDefaultView();
+}
+
+::dawn::ShaderModule CreateShaderModuleFromResult(
+    const ::dawn::Device& device,
+    const shaderc::SpvCompilationResult& result) {
+  // result.cend and result.cbegin return pointers to uint32_t.
+  const uint32_t* resultBegin = result.cbegin();
+  const uint32_t* resultEnd = result.cend();
+  // So this size is in units of sizeof(uint32_t).
+  ptrdiff_t resultSize = resultEnd - resultBegin;
+  // SetSource takes data as uint32_t*.
+
+  ::dawn::ShaderModuleDescriptor descriptor;
+  descriptor.codeSize = static_cast<uint32_t>(resultSize);
+  descriptor.code = result.cbegin();
+  return device.CreateShaderModule(&descriptor);
+}
+
+shaderc_shader_kind ShadercShaderKind(::dawn::ShaderStage stage) {
+  switch (stage) {
+    case ::dawn::ShaderStage::Vertex:
+      return shaderc_glsl_vertex_shader;
+    case ::dawn::ShaderStage::Fragment:
+      return shaderc_glsl_fragment_shader;
+    case ::dawn::ShaderStage::Compute:
+      return shaderc_glsl_compute_shader;
+    default:
+      return shaderc_glsl_vertex_shader;
+  }
+}
+
+::dawn::ShaderModule CreateShaderModule(const ::dawn::Device& device,
+                                        ::dawn::ShaderStage stage,
+                                        const char* source) {
+  shaderc_shader_kind kind = ShadercShaderKind(stage);
+
+  shaderc::Compiler compiler;
```

Comment:
This should not need to call shaderc, the shader module is already compiled by the time we create the framework. The data is in the shader info.